openmmtools/integrators.py

```diff
@@ -671,6 +672,206 @@ def __init__(self, temperature=298 * simtk.unit.kelvin, collision_rate=91.0 / si
+
+        timestep: simtk.unit.Quantity compatible with femtoseconds, default=1*simtk.unit.femtoseconds
+            The integration timestep for particles.
+
+        chain_length: integer, default=5
+            The number of thermostat particles in the Nosé-Hoover chain.  Increasing
+            this parameter will affect computational cost, but will make the simulation
+            less sensitive to thermostat parameters, particularly in stiff systems; see
+            the 1992 paper referenced above for more details.
+
+        num_mts: integer, default=5
+            The number of timesteps used in the multi-timestep procedure to update the thermostat
+            positions.  A higher value will increase the stability of the dynamics, but will also
+            increase the compuational cost of the integration.
+
+        num_yoshidasuzuki: integer, default=5
+            The number of Yoshida-Suzuki steps used to subdivide each of the multi-timesteps used
+            to update the thermostat positions.  A higher value will increase the stability of the
+            dynamics, but will also increase the computational cost of the integration; only certain
+            values (currently 1,3, or 5) are supported, because weights must be defined.
+
+        """
+        super(NoseHooverChainVelocityVerletIntegrator, self).__init__(temperature, timestep)
+        #
+        # Integrator initialization.
+        #
+        self.n_c         = num_mts
+        self.n_ys        = num_yoshidasuzuki
+        try:
+            self.weights     = self.YSWeights[self.n_ys]
+        except KeyError:
+            raise Exception("Invalid Yoshida-Suzuki value. Allowed values are: %s"%
+                             ",".join(map(str,self.YSWeights.keys())))
+        if chain_length < 0:
+            raise Exception("Nosé-Hoover chain length must be at least 0")
+        if chain_length == 0:
+            print("WARNING: Nosé-Hoover chain length is 0; falling back to regular velocity verlet algorithm.")
+        self.M           = chain_length
+
+        # Define the "mass" of the thermostat particles (multiply by ndf for particle 0)
+        kT = self.getGlobalVariableByName('kT')
+        frequency = collision_frequency.value_in_unit(simtk.unit.picoseconds**-1)
+        Q = kT/frequency**2
+
+        #
+        # Define global variables
+        #
+        self.addGlobalVariable("ndf", 0)      # number of degrees of freedom
+        self.addGlobalVariable("bathKE", 0.0) # Thermostat bath kinetic energy
+        self.addGlobalVariable("bathPE", 0.0) # Thermostat bath potential energy
+        self.addGlobalVariable("KE2", 0.0)    # Twice the kinetic energy
+        self.addGlobalVariable("Q", Q)        # Thermostat particle "mass"
+        self.addGlobalVariable("scale", 1.0)
+        self.addGlobalVariable("aa", 0.0)
+        self.addGlobalVariable("wdt", 0.0)
+        for w in range(self.n_ys):
+            self.addGlobalVariable("w%d"%w, self.weights[w])
```

Comment:
Small consistency thing, can you switch all your string formatters to `.format` over `%`?